### ISSUE TITLE
dask-ssh will not send correct address when scheduler is localhost

### ISSUE THREAD

**Issue by landonzweigle** [dask/dask#5087]
Hello (again!),
When using dask-ssh specifying localhost as the scheduler, "localhost" is the address given to the workers, so no connection is made if the workers aren't on localhost as well. Easy solution is not specifying scheduler as localhost but as the "actual" address. Still a problem in my opinion, though! The console will repeatedly say:

```[ worker NotARealMachine.edu ] : distributed.worker - INFO - Waiting to connect to: tcp://localhost:8786```

reproduce:
`dask-ssh --scheduler localhost NotARealMachine.edu AnotherFakemachine.edu`
This occurs for both TCP and TLS connections.

**Comment by quasiben** [comment]
I am not sure this is a bug.  If the scheduler is not addressable by remote workers then there is not much dask can do.  This would be true if using `localhost` or `127.0.0.1`.  What would you expect dask to do differently here ?

**Comment by landonzweigle** [comment]
I see. I was thinking localhost could be resolved to a public address, such as the address given by if/ipconfig. Getting an actual public address might require a call to some external API or something. Sorry if I am wrong about this, I am not super familiar with networking and as such didn't give it much thought other than "huh, that doesn't seem right".

**Comment by quasiben** [comment]
@landonzweigle , I agree, this is annoying when you want to resolve localhost to the *most* available address.  This is somewhat challenging in many if not most cases -- my machine right now has a few IPs: 127.0.0.1, internal ip 192.168.xxx, and it's also addressable because of router configurations allowing connections on XXXXX for the global address for my office/home/etc.  You can see this with `socket`:

```python
import socket
socket.gethostbyname(socket.gethostname()) # will probably return 127.0.0.1

import socket
s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
s.connect(("8.8.8.8", 80))
print(s.getsockname()[0]) # probably internal ip
s.close()
```

You can also see all this info with `ifconfig or /sbin/ifconfig` (if on OSX/Linux).  The problem may even be worse with several attached network cards or other devices which expose network interfaces.

So with all that said, many systems offload this problem to the user and let them define where the connection should happen (locahost, private ip, global address, k8s internal address, etc).  Still, I agree this is a rather annoying problem which takes away 10s of minutes to hours from users when trying to debug unfamiliar networking issues.

**Comment by landonzweigle** [comment]
> Still, I agree this is a rather annoying problem which takes away 10s of minutes to hours from users when trying to debug unfamiliar networking issues.

I was thinking simply warn the user if the address can't be resolved (or something...). I was also thinking that if localhost is used, tell the workers that the connection that was used to initialize them is the same connection as the scheduler (because some communication with "localhost" had to be made, right?)... I really have no idea how you do things with setting up workers, so maybe this wouldn't work or would take longer to implement as a simple external api (ipify).

**Comment by quasiben** [comment]
@jcrist do you have thoughts here ?  My opinion is that while the log comment:

> Waiting to connect to: tcp://localhost:8786

is not an explicit comment saying unable to resolve host a user could (should?) infer it ?

**Comment by jcrist** [comment]
> I was thinking localhost could be resolved to a public address, such as the address given by if/ipconfig.

While we could make an attempt at this, it's not the safest approach. What if the user really meant localhost rather than something publically addressable? What if they use another loopback address? I'd rather not guess the user's intent, and fix this with documentation/better logging.

**Comment by jrbourbeau** [comment]
> fix this with documentation/better logging

@landonzweigle do you an interest in submitting a PR? (no obligation though)

**Comment by quasiben** [comment]
@landonzweigle one possibility is that after N attempts the message could change from:
> Waiting to connect to: tcp://localhost:8786

to something more informative/exploratory:

> Connection to: tcp://localhost:8786 is still waiting -- user should confirm host is addressable by worker 

@landonzweigle, I would also say that error messages can be difficult to craft.  Thank you for raising this issue and it's more than likely you are not the only person to have experienced this problem
